### DISCUSSION TITLE
Puma, localhost, Swift, URLSession, timeout errors

### DISCUSSION THREAD

**Discussion by groue** [puma/puma#3394] [Q&A]
**Describe the bug**

Hello,

When I run my end-to-end tests from Swift against my localhost Ruby server (Puma, Rails 7), I sometimes get network timeout errors (Foundation `URLError` code -1001).

Server looks perfectly fine. It's just as if the client does not receive some replies.

I never have those errors when the same tests run against a remote server (nginx+passenger).

*I'm not sure this is a problem with Puma.* I'm mostly looking for advice and/or previous experience. Does this ring a bell to you?

**Puma config:**

It is the default Puma config installed by Rails 7. I'm not sure where to find it.

<details><summary>config/puma.rb</summary>

```ruby
# Puma can serve each request in a thread from an internal thread pool.
# The `threads` method setting takes two numbers: a minimum and maximum.
# Any libraries that use thread pools should be configured to match
# the maximum value specified for Puma. Default is set to 5 threads for minimum
# and maximum; this matches the default thread size of Active Record.
#
max_threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }
min_threads_count = ENV.fetch("RAILS_MIN_THREADS") { max_threads_count }
threads min_threads_count, max_threads_count

# Specifies the `worker_timeout` threshold that Puma will use to wait before
# terminating a worker in development environments.
#
worker_timeout 3600 if ENV.fetch("RAILS_ENV", "development") == "development"

# Specifies the `port` that Puma will listen on to receive requests; default is 3000.
#
port ENV.fetch("PORT") { 3000 }

# Specifies the `environment` that Puma will run in.
#
environment ENV.fetch("RAILS_ENV") { "development" }

# Specifies the `pidfile` that Puma will use.
pidfile ENV.fetch("PIDFILE") { "tmp/pids/server.pid" }

# Specifies the number of `workers` to boot in clustered mode.
# Workers are forked web server processes. If using threads and workers together
# the concurrency of the application would be max `threads` * `workers`.
# Workers do not work on JRuby or Windows (both of which do not support
# processes).
#
# workers ENV.fetch("WEB_CONCURRENCY") { 2 }

# Use the `preload_app!` method when specifying a `workers` number.
# This directive tells Puma to first boot the application and load code
# before forking the application. This takes advantage of Copy On Write
# process behavior so workers use less memory.
#
# preload_app!

# Allow puma to be restarted by `bin/rails restart` command.
plugin :tmp_restart
```

</details>

```
$ bin/rails s
=> Booting Puma
=> Rails 7.0.8 application starting in development 
=> Run `bin/rails server --help` for more startup options
Puma starting in single mode...
* Puma version: 5.6.7 (ruby 3.2.2-p53) ("Birdie's Version")
*  Min threads: 5
*  Max threads: 5
*  Environment: development
* Listening on http://127.0.0.1:3000
* Listening on http://[::1]:3000
Use Ctrl-C to stop
```

**To Reproduce**

Run many requests (serial or parallel) until the client (`URLSession`) starts to fail with timeout errors (in some random request). This typically happens in end-to-end tests.

Adding a delay (up to 1s) between requests does not fix the problem.

**Expected behavior**

No timeout errors in the client.

**Desktop (please complete the following information):**
 - OS: MacOS 14.3.1 (23D60)
 - Puma version: 5.6.7 (ruby 3.2.2-p53) ("Birdie's Version")

---

:thinking: If Puma is HTTP 1.1, and nginx HTTP 2, maybe the difference that matters is not localhost+Puma vs. remote+nginx/passenger, it would be the HTTP version. In which case the real problem could lie in the Swift `URLSession` component which could badly handle HTTP 1.1, however unlikely it is... I'm a little bit lost, and don't quite know where to ask for support.

**Comment by dentarg** [comment]
I converted your issue to a discussion. Let's open an issue when (if) we know there's something actionable in Puma.

You will need to share more details (how many is 'many' requests, for example) and code so others can try to reproduce the issue, to understand what's happening. 

It could be that the system is out of resources (memory, file handles/descriptors). (Just a guess, as stated, not enough details to say)

**Reply by dentarg** [reply]
Please also try with the latest version of Puma, or even the main branch.

**Reply by groue** [reply]
Thank you @dentarg, yes I will try to upgrade Puma and check again.

For information, this problem has nothing to do with the server load. I also have client errors:
- when I serialize requests (no concurrent server access)
- when I put a generous delay between requests (up to 1 second)

The timeout errors happen after a while. If I restart my tests, without restarting the server, I witness the same behavior (ok for a while, until errors start to happen, but never at the same exact moment). This might sound like an issue in the client HTTP stack, yes. Maybe even an exhaustion of file handle descriptors, as you say. Or some weird interaction with Puma. On that topic, I was [nicely suggested](https://hachyderm.io/@YOCKOW@mastodon.yockow.jp/112493001325196673) to use a different HTTP client stack. After all, maybe the Apple-provided stack has bugs.

These errors never happen when I run the same tests against a remote server, and that's why I came here after all.

Anyway. I do not accuse anyone.

I'll try to take the time to do all those checks, and report my findings.

If you read this, and had a similar issue, please share your experience!
